adr/ADR-32.md

```diff
@@ -132,24 +132,129 @@ un-acked.
+
+Users should be able to set either max_messages or max_bytes values, but not both:
+
+- If no option is provided, the default value for `max_messages` (1000) should be used, and
+`max_bytes` should not be set.
+- If `max_messages` is set by the user, the value should be set for `max_messages` and `max_bytes`
+should not be set.
+- User cannot set both constraint for a single `Consume()` execution.
+- For each constraint, a custom threshold can be set, containing the number of messages/bytes that should be received to
+trigger the next pull request. The value of threshold cannot be higher than the corresponding constraint's value.
+- Default values for threshold should be 25% of `max_messages`/`max_bytes`
+- For each pull request, `batch` or `max_bytes` value should be equal to the threshold value (to fill the buffer)
+
+###### Buffering messages
+
+`Consume()` should pre-buffer messages up to a limit set by `max_messages` or `max_bytes` options (whichever is provided).
+Clients should track the total amount of messages pending in a buffer. Whenever a threshold is reached,
+a new request to `CONSUMER.MSG.NEXT` should be published.
+
+There is no need to track specific pull request's status - as long as the aggregate message and byte count is
+maintained, `Consume()` should be able to fill the buffer appropriately.
+
+Pending messages and bytes count should be updated when:
+
+- A new pull request is published - add a value of `request.batch_size` to the pending messages count and
+the value of `request.max_bytes` to the pending byte count.
+- A new user message is processed - subtract 1 from pending messages count and subtract message size from penging byte count.
+The message size (in bytes) is defined as: `len(msg.data) + len(msg.subject) + len(msg.reply)`
+- A pull request termination error is received - subtract the value of `Nats-Pending-Messages` header
+from pending messages count and subtract the value of `Nats-Pending-Bytes` from pending bytes count.
+
+###### JetStream error handling
+
+There are 2 kinds of errors which can be received as a response to a pull request:
+
+- Pull request termination error is any error containing `Nats-Pending-Messages` and `Nats-Pending-Bytes` headers.
+These should not be treated as errors by the application, but rather used to calculate the pending messages/bytes count.
+- Every other pull request error should be treated as terminal error - it should be telegraphed to the user (in langie-specific way),
+the `Consume()` execution should be stopped and subscription should be drained.
+
+###### Idle heartbeats
+
+`Consume()` should always utilize idle heartbeats. Heartbeat values are calculated as follows:
+
+- If no heartbeat value is provided by the user:
+  - if `expiry < 60s`, set the heartbeat value to `expiry / 2`
+  - if `expiry >= 60`, set the heartbeat value to 30s
+- If heartbeat value is provided by the user, validate whether `idle_heartbeat > 500ms && idle_heartbeat <= expires / 2`
```

Comment:
This could still leed to some very large amount of heartbeats being sent, is there really value in heartbeats more frequently than 10 or even 30 seconds?

Reply:
We discussed this with @Jarema and apparently we might have some users who value high responsiveness. Although I agree that this should generally be discouraged.

The other issue here is, if we allow users to set very low expiry (e.g. 1s, question is if this is even a valid use case), we need to set heartbeats which are <= expiry / 2 and we decided that we always want to utilize heartbeats in `Consume()`, to avoid potentially being stuck forever.

Reply:
Well an expiry 1 consumer won’t be stuck forever will it? It will be stuck for 1 second. 

We need to consider the cost on the server - especially big servers - when we allow things to be tbat aggressive. Just because you can have 100 000 clients each with sub 1 second heartbeats doesn’t mean you should

Reply:
`Consume()` endlessly fetches messages from the server - so in that case `expires` setting is only really used to fine-tune / optimize the execution (how many requests we'll be sending if we do not have steady flow of messages). If a request times out on expiry, that only means that the next pull request will be published to attempt fetching more messages.

To your point, yes, we should consider and document this well. However, notice that the defaults are far more sensible (30s expiry, 15s heartbeats). So while I would assume that users might often play around with `max_messages` or `max_bytes` options (to optimize their performance and memory footprint), expiry and heartbeat settings are targeted towards "power users" who really know what they're doing. I believe that if this is well documented _in each client_, it might be valuable to at least some small numbers of users.

Reply:
Example of adding a new service looks sth like this:
```go
	nc, err := nats.Connect("127.0.0.1:4222")
	if err != nil {
		log.Fatal(err)
	}

	config := Config{
		Name:     "EchoService",
		Endpoint: Endpoint{Subject: "echo", Handler: echoHandler},
	}

	srv, err := service.Add(nc, config)
	if err != nil {
		log.Fatal(err)
	}
```